src/Features/LanguageServer/Protocol/Handler/SpellCheck/AbstractSpellCheckingHandler.cs

```diff
@@ -0,0 +1,292 @@
+            // last time we notified the client.  Report back either to the client so they can update accordingly.
+            var orderedDocuments = GetOrderedDocuments(context, cancellationToken);
+            context.TraceInformation($"Processing {orderedDocuments.Length} documents");
+
+            foreach (var document in orderedDocuments)
+            {
+                context.TraceInformation($"Processing: {document.FilePath}");
+
+                var languageService = document.GetLanguageService<ISpellCheckSpanService>();
+                if (languageService == null)
+                {
+                    context.TraceInformation($"Ignoring document '{document.FilePath}' because it does not support spell checking");
+                    continue;
+                }
+
+                if (!IncludeDocument(document, context.ClientName))
+                {
+                    context.TraceInformation($"Ignoring document '{document.FilePath}' because of razor/client-name mismatch");
+                    continue;
+                }
+
+                var newResultId = await GetNewResultIdAsync(documentToPreviousParams, document, cancellationToken).ConfigureAwait(false);
+                if (newResultId != null)
+                {
+                    context.TraceInformation($"Spans were changed for document: {document.FilePath}");
+                    progress.Report(await ComputeAndReportCurrentSpansAsync(
+                        document, languageService, newResultId, cancellationToken).ConfigureAwait(false));
+                }
+                else
+                {
+                    context.TraceInformation($"Spans were unchanged for document: {document.FilePath}");
+
+                    // Nothing changed between the last request and this one.  Report a (null-spans, same-result-id)
+                    // response to the client as that means they should just preserve the current spans they have for
+                    // this file.
+                    var previousParams = documentToPreviousParams[document];
+                    progress.Report(CreateReport(previousParams.TextDocument, ranges: null, previousParams.PreviousResultId));
+                }
+            }
+
+            // If we had a progress object, then we will have been reporting to that.  Otherwise, take what we've been
+            // collecting and return that.
+            context.TraceInformation($"{this.GetType()} finished getting spans");
+            return progress.GetValues();
+        }
+
+        private static bool IncludeDocument(Document document, string? clientName)
+        {
+            // Documents either belong to Razor or not.  We can determine this by checking if the doc has a span-mapping
+            // service or not.  If we're not in razor, we do not include razor docs.  If we are in razor, we only
+            // include razor docs.
```

Comment:
this applies to diagnostics as well, but it might make more sense for this to move into the workspace pull handler.  Document diagnostics requests should only ever come into the appropriate server.  Only Razor should be asking for doc diagnostics for a razor file.